hazelcast-sql/src/main/codegen/includes/parserImpls.ftl

```diff
@@ -0,0 +1,312 @@
+    |
+        <SMALLINT> { type = QueryDataType.SMALLINT; }
+    |
+        (<INTEGER> | <INT>) { type = QueryDataType.INT; }
+    |
+        <BIGINT> { type = QueryDataType.BIGINT; }
+    |
+        (<REAL> | <FLOAT>) { type = QueryDataType.REAL; }
+    |
+        <DOUBLE> [ <PRECISION> ] { type = QueryDataType.DOUBLE; }
+    |
+        (<DECIMAL> | <DEC> | <NUMERIC>)
+        [
+            <LPAREN>
+            precision = UnsignedIntLiteral()
+            [
+                <COMMA>
+                scale = UnsignedIntLiteral()
+            ]
+            <RPAREN>
+        ] { type = scale > 0 ? QueryDataType.DECIMAL : QueryDataType.DECIMAL_BIG_INTEGER; }
+    )
+    {
+        return type;
+    }
+}
+
+QueryDataType CharacterType() :
+{
+    QueryDataType type;
+}
+{
+    (
+        (<CHARACTER> | <CHAR>)
+        (
+            <VARYING> { type = QueryDataType.VARCHAR; }
+        |
+            { type = QueryDataType.VARCHAR_CHARACTER; }
+        )
+    |
+        <VARCHAR> { type = QueryDataType.VARCHAR; }
+    )
+    {
+        return type;
+    }
+}
+
+QueryDataType DateTimeType() :
+{
+    QueryDataType type;
+    SqlIdentifier variant = null;
+}
+{
+    (
+        <TIME> { type = QueryDataType.TIME; }
+    |
+        <DATE> { type = QueryDataType.DATE; }
+    |
+        <TIMESTAMP>
+        (
+            <WITH>
+            (
+                <TIME> <ZONE>
+                [
+                    <LPAREN> variant = SimpleIdentifier() <RPAREN>
+                ]
+                {
+                    if (variant == null) {
+                        type = QueryDataType.TIMESTAMP_WITH_TZ_OFFSET_DATE_TIME;
+                    } else if ("ZONED_DATE_TIME".equalsIgnoreCase(variant.getSimple())) {
+                        type = QueryDataType.TIMESTAMP_WITH_TZ_ZONED_DATE_TIME;
+                    } else if ("CALENDAR".equalsIgnoreCase(variant.getSimple())) {
+                        type = QueryDataType.TIMESTAMP_WITH_TZ_CALENDAR;
+                    } else {
+                        throw SqlUtil.newContextException(getPos(),
+                            ParserResource.RESOURCE.unknownTimestampVariant(variant.getSimple()));
+                    }
+                }
+            |
+                <LOCAL> <TIME> <ZONE>
+                [
+                    <LPAREN> variant = SimpleIdentifier() <RPAREN>
+                ]
+                {
+                    if (variant == null) {
+                        type = QueryDataType.TIMESTAMP_WITH_TZ_INSTANT;
+                    } else if ("DATE".equalsIgnoreCase(variant.getSimple())) {
+                        type = QueryDataType.TIMESTAMP_WITH_TZ_DATE;
+                    } else {
+                        throw SqlUtil.newContextException(getPos(),
+                            ParserResource.RESOURCE.unknownTimestampVariant(variant.getSimple()));
+                    }
+                }
+            )
+        |
+            <WITHOUT> <TIME> <ZONE> { type = QueryDataType.TIMESTAMP; }
```

Comment:
This should be also the default when just `TIMESTAMP` is used without any TIME ZONE clause